sandbox/pc2sandbox.sh

```diff
@@ -0,0 +1,234 @@
+###MEMLIMIT=8
+###TIMELIMIT=2
+###COMMAND=$1
+###shift
+
+# the rest is the command args
+
+# make sure we have CGroups V2 properly installed on this system, including a PC2 structure
+
+DEBUG echo checking PC2 CGroup V2 installation...
+if [ ! -d "$PC2_CGROUP_PATH" ]; then
+   echo $0: expected pc2sandbox CGroups v2 installation in $PC2_CGROUP_PATH 
+   exit $FAIL_INVALID_CGROUP_INSTALLATION
+fi
+
+if [ ! -f "$CGROUP_PATH/cgroup.controllers" ]; then
+   echo $0: missing file cgroup.controllers in $CGROUP_PATH
+   exit $FAIL_MISSING_CGROUP_CONTROLLERS_FILE
+fi
+
+if [ ! -f "$CGROUP_PATH/cgroup.subtree_control" ]; then
+   echo $0: missing file cgroup.subtree_control in $CGROUP_PATH
+   exit $FAIL_MISSING_CGROUP_SUBTREE_CONTROL_FILE
+fi
+
+# make sure the cpu and memory controllers are enabled
+if ! grep -q -F "cpu" "$CGROUP_PATH/cgroup.subtree_control"; then
+   echo $0: cgroup.subtree_control in $CGROUP_PATH does not enable cpu controller
+   exit $FAIL_CPU_CONTROLLER_NOT_ENABLED
+fi
+
+if ! grep -q -F "memory" "$CGROUP_PATH/cgroup.subtree_control"; then
+   echo $0: cgroup.subtree_control in $CGROUP_PATH does not enable memory controller
+   exit $FAIL_MEMORY_CONTROLLER_NOT_ENABLED
+fi
+
+
+# we seem to have a valid CGroup installation
+DEBUG echo ...done.
+
+if test -d $PC2_SANDBOX_CGROUP_PATH
+then
+	DEBUG echo Removing existing sandbox to start clean
+	rmdir $PC2_SANDBOX_CGROUP_PATH
+fi
+
+DEBUG echo Creating sandbox $PC2_SANDBOX_CGROUP_PATH
+if ! mkdir $PC2_SANDBOX_CGROUP_PATH
+then
+	DEBUG echo Can not create $PC2_SANDBOX_CGROUP_PATH
+	exit $FAIL_INVALID_CGROUP_INSTALLATION
+fi
+
+# set the specified memory limit - input is in MB, cgroup v2 requires bytes, so multiply by 1M
+# but only if > 0.
+# "max" means unlimited, which is the cgroup v2 default
+DEBUG echo checking memory limit
+if [ "$MEMLIMIT" -gt "0" ] ; then
+  DEBUG echo setting memory limit to $MEMLIMIT MB
+  echo $(( $MEMLIMIT * 1024 * 1024 ))  > $PC2_SANDBOX_CGROUP_PATH/memory.max
+  echo 1  > $PC2_SANDBOX_CGROUP_PATH/memory.swap.max
+else
+  DEBUG echo setting memory limit to max, meaning no limit
+  echo "max" > $PC_SANDBOX_CGROUP_PATH/memory.max  
+  echo "max" > $PC_SANDBOX_CGROUP_PATH/memory.swap.max  
+fi
+
+# set the specified CPU time limit - input is in secs, cgroup v2 requires usec, so multiply by 1M.
+# cgroup v2 expects two parameters:  absolute time and "period", but if only one is provided it is "absolute time"
+DEBUG echo setting cpu limit to $TIMELIMIT seconds
+ulimit -t $TIMELIMIT
+TIMELIMIT_US=$((TIMELIMIT * 1000000))
+#echo $TIMELIMIT_US  > $PC2_SANDBOX_RUN_CGROUP_PATH/cpu.max
+
+DEBUG echo setting maximum user processes to 32 + whatever the user is currently using
+ulimit -u $((32+`ps -T -u $USER | wc -l`))
+
+#put the current process (and implicitly its children) into the pc2sandbox cgroup.
+DEBUG echo putting $$ into $PC2_SANDBOX_CGROUP_PATH cgroup
+if ! echo $$ > $PC2_SANDBOX_CGROUP_PATH/cgroup.procs
+then
+	echo Failed. quitting.
```

Comment:
If this happens, all the user will see is "Failed. quitting." with no explanation.  It's true there is additional information in the DEBUG (log) file -- but only if _DEBUG has remained set to "on", and even then it's only useful if the user knows where to look.  I think the message should at _least_ contain a little bit of help, like `echo Failed to add process to CGroups process list. Quitting` or something similar.

Reply:
In fact, I don't think that message will be seen by anyone anyway.  The fact it returns 1 will cause pc2 to give an error with a return code of 1, and then one can look at the script to see what happened.  I think I will change it to be a "well-known" return value like all the FAIL_XXX at the top (FAIL_CGROUP_PID_ADD?), this way it is easier to see what happened, and those will generate an OCS in some cases.  See a little further down in the script you had:

# TODO: determine how to pass pc2sandbox.sh results back to PC2...

And that would be a neat trick - to get detailed results back and report them.  Right now, it just goes by the one byte exit code.

Reply:
Fixed.  Added better message and new return fail code.